internal/events/first_eventbatch_preparer.go

```diff
@@ -0,0 +1,105 @@
+package events
+
+import (
+	"context"
+	"fmt"
+
+	apiv1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . FirstEventBatchPreparer
+
+// FirstEventBatchPreparer prepares the first batch of events to be processed by the EventHandler.
+// The first batch includes the UpsertEvents for all relevant resources in the cluster.
+type FirstEventBatchPreparer interface {
+	// Prepare prepares the first event batch.
+	Prepare(ctx context.Context) (EventBatch, error)
+}
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . CachedReader
+
+// CachedReader allows getting and listing resources from a cache.
+// This interface is introduced for testing to mock a subset of methods from
+// sigs.k8s.io/controller-runtime/pkg/cache.Cache.
+type CachedReader interface {
+	WaitForCacheSync(ctx context.Context) bool
+	Get(ctx context.Context, key client.ObjectKey, obj client.Object) error
+	List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error
+}
+
+// FirstEventBatchPreparerImpl is an implementation of FirstEventBatchPreparer.
+type FirstEventBatchPreparerImpl struct {
+	reader CachedReader
+	gcName string
+}
+
+// NewFirstEventBatchPreparerImpl creates a new FirstEventBatchPreparerImpl.
+func NewFirstEventBatchPreparerImpl(reader CachedReader, gcName string) *FirstEventBatchPreparerImpl {
+	return &FirstEventBatchPreparerImpl{
+		reader: reader,
+		gcName: gcName,
+	}
+}
+
+func (p *FirstEventBatchPreparerImpl) Prepare(ctx context.Context) (EventBatch, error) {
+	synced := p.reader.WaitForCacheSync(ctx)
+	if !synced {
+		return nil, fmt.Errorf("cache is not synced")
+	}
+
+	var gc v1beta1.GatewayClass
+	gcExist := true
+	err := p.reader.Get(ctx, types.NamespacedName{Name: p.gcName}, &gc)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			gcExist = false
+		} else {
+			return nil, err
+		}
+	}
+
+	var svcList apiv1.ServiceList
+	var secretList apiv1.SecretList
+	var gwList v1beta1.GatewayList
+	var hrList v1beta1.HTTPRouteList
+
+	objLists := []client.ObjectList{&svcList, &secretList, &gwList, &hrList}
```

Comment:
I think an improved API is to either construct the Preparer with a list or pass it as an argument here.

Passing in the list allows for flexible testing. Passing in the list simplifies future updates.

Reply:
Not sure if there is a type I can use to pass those lists.

In `err := p.reader.List(ctx, list)`, we pass list as  `client.ObjectList`, which doesn't have methods to access the Items array. However, we still need to access items as in:
```
	for i := range svcList.Items {
		batch = append(batch, &UpsertEvent{Resource: &svcList.Items[i]})
	}
```
So if I pass multiple `client.ObjectList` to `Prepare()`, I will still need to cast it to a typed list like `apiv1.ServiceList`.

Reply:
I think using help.go from controller-runtime will work here.

Prepare takes a list:
`client.ObjectList{&svcList, &secretList}`

The for loop at 71 doesn't change.

Instead of line 83 it becomes something effective to (using range and proper loop):
`batch := make([]interface, 0, gcCount+LenList(objLists[0])+LenList(objLists[1])` 

At 91:
```
for each objList
  EachListItem(list, fn)
```

where fn:
```
func (runtime.Object) error {
  switch obj := runtime.Object.(type) {
  case client.Object:
    // append to batch
  }
}
```